lib/kubernetes-deploy/discoverable_resource.rb

```diff
@@ -0,0 +1,177 @@
+# frozen_string_literal: true
+require 'kubernetes-deploy/kubernetes_resource'
+require 'kubernetes-deploy/kubeclient_builder'
+require "pry"
+require 'jsonpath'
+
+module KubernetesDeploy
+  class DiscoverableResource < KubernetesResource
+    extend KubernetesDeploy::KubeclientBuilder
+
+    class << self
+      attr_accessor :version, :group, :type
+    end
+
+    STATUS_FIELD_ANNOTATION = 'kubernetes-deploy.shopify.io/status-field'
+    STATUS_SUCCESS_ANNOTATION = 'kubernetes-deploy.shopify.io/status-success'
+    TIMEOUT_ANNOTATION = 'kubernetes-deploy.shopify.io/timeout'
+    PREDEPLOY_ANNOTATION = 'kubernetes-deploy.shopify.io/predeploy'
+    PRUNABLE_ANNOTATION = 'kubernetes-deploy.shopify.io/prunable'
+
+    def type
+      self.class.type
+    end
+
+    def self.inherited(child_class)
+      child_classes.add child_class
+    end
+
+    def self.child_classes
+      @child_classes ||= Set.new
+    end
+
+    def self.prunable?
+      return @prunable if defined?(@prunable)
+      false
+    end
+
+    def self.predeploy?
+      return @predeploy if defined?(@predeploy)
+      false
+    end
+
+    def self.timeout
+      return @timeout if defined?(@timeout)
+      super
+    end
+
+    def self.identity
+      "#{group}/#{version}/#{type}"
+    end
+
+    def self.all
+      child_classes.dup
+    end
+
+    def self.discover(context:, logger:)
+      logger.info("Discovering custom resources:")
+      @resources = nil
+      @child_classes = nil
+      discover_tpr(v1beta1_kubeclient(context))
+      begin
+        discover_crd(v1beta1_crd_kubeclient(context))
+      rescue KubeException => err
+        logger.warn("Unable to discover CustomResourceDefinitions: #{err}")
+      end
+    end
+
+    def self.build(namespace:, context:, definition:, logger:)
+      return super if KubernetesDeploy.const_defined?(definition["kind"])
```

Comment:
This allows us to temporarily fall back to the custom classes in `k8s-deploy` if they are present for a given resource. This behaviour is useful for progressively moving the monitoring to the controllers and outside of this gem.

Reply:
(irrelevant if we can make the instances direct subclasses of KubernetesResource, but...) Doesn't this end up covering the regular resources, i.e. is not temporary?

Reply:
This is more explicit in the re-written version. If a matching (== kind) symbol exists in the k8s-deploy module it is used, otherwise a class is generated dynamically.